Implement WebAuthn Signal API

pascoej opened the pull request:
#### 445bbfd2a3de74ce9b57ebbd2ed0d0271c473e5c
<pre>
Implement WebAuthn Signal API
<a href="https://bugs.webkit.org/show_bug.cgi?id=278339">https://bugs.webkit.org/show_bug.cgi?id=278339</a>
<a href="https://rdar.apple.com/problem/134285651">rdar://problem/134285651</a>

Reviewed by Nitin Mahendru.

This patch implements the new WebAuthn signal API. This allows RPs to
tell authenticators what credentials it will accept and also to update
credential information.

We don&apos;t currently mock the information nessesary to write tests for these
methods in LayoutTests.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/webauthn/AllAcceptedCredentialsOptions.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredential.idl.
* Source/WebCore/Modules/webauthn/AllAcceptedCredentialsOptions.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredential.idl.
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::signalUnknownCredential):
(WebCore::AuthenticatorCoordinator::signalAllAcceptedCredentials):
(WebCore::AuthenticatorCoordinator::signalCurrentUserDetails):
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.h:
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinatorClient.h:
* Source/WebCore/Modules/webauthn/CurrentUserDetailsOptions.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredential.idl.
* Source/WebCore/Modules/webauthn/CurrentUserDetailsOptions.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredential.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredential.cpp:
(WebCore::PublicKeyCredential::signalUnknownCredential):
(WebCore::PublicKeyCredential::signalAllAcceptedCredentials):
(WebCore::PublicKeyCredential::signalCurrentUserDetails):
* Source/WebCore/Modules/webauthn/PublicKeyCredential.h:
* Source/WebCore/Modules/webauthn/PublicKeyCredential.idl:
* Source/WebCore/Modules/webauthn/UnknownCredentialOptions.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredential.idl.
* Source/WebCore/Modules/webauthn/UnknownCredentialOptions.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredential.idl.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.h:
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/AuthenticationServices/CredentialUpdaterShim.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredential.idl.
* Source/WebKit/WebKitSwift/AuthenticationServices/CredentialUpdaterShim.swift: Added.
(CredentialUpdaterShim.signalUnknownCredential(_:credentialID:)):
(CredentialUpdaterShim.signalAllAcceptedCredentials(_:userHandle:acceptedCredentialIDs:)):
(CredentialUpdaterShim.signalCurrentUserDetails(_:userHandle:newName:)):
* Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.cpp:
(WebKit::WebAuthenticatorCoordinator::signalUnknownCredential):
(WebKit::WebAuthenticatorCoordinator::signalAllAcceptedCredentials):
(WebKit::WebAuthenticatorCoordinator::signalCurrentUserDetails):
* Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.h:

Canonical link: <a href="https://commits.webkit.org/291458@main">https://commits.webkit.org/291458@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7201075b68efd974d18ff4af2ae9e382d2d3284

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92962 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12513 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2176 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97961 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43488 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95012 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12794 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20965 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71077 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28494 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95964 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9637 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84090 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51405 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9330 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1726 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42801 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79622 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1700 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99983 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20014 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14665 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80100 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20265 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79989 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79401 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19718 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23947 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1220 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13046 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19998 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25174 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19685 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23145 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21426 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/664bc3ebf77194dfedd7ddb6bc48def7a5c84956. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/41630

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/04142588e8773e336e8022f3c869050239a87986. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/41630

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c4bc0964a31d32286fc4ec7cfee8330822576d48. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/41630

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/16212a8c27c5fa82bfeef621e1f9a36042089f72. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/41630

nmahendru commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -424,6 +424,42 @@ void AuthenticatorCoordinator::getClientCapabilities(const Document& document, D
+void AuthenticatorCoordinator::signalUnknownCredential(const Document& document, UnknownCredentialOptions&& options, DOMPromiseDeferred<void>&& promise)
+{
+    if (!m_client)  {
+        promise.reject(Exception { ExceptionCode::UnknownError, "Unknown internal error."_s });
```
log or a better exception message ?

nmahendru commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.h:
```diff
@@ -60,7 +61,6 @@ struct PublicKeyCredentialRequestOptions;
 struct CredentialCreationOptions;
 struct CredentialRequestOptions;
 class SecurityOriginData;
-
```
nit: This line can probably stay.

nmahendru commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -1281,6 +1287,44 @@ static inline void getArePasskeysDisallowedForRelyingParty(const WebCore::Securi
 
+void WebAuthenticatorCoordinatorProxy::signalUnknownCredential(const WebCore::SecurityOriginData&, WebCore::UnknownCredentialOptions&& options, CompletionHandler<void()>&& completionHandler)
+{
+    auto decodedCredentialId = base64URLDecode(options.credentialId);
```
Error checking for URLDecode ?

nmahendru commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -1281,6 +1287,44 @@ static inline void getArePasskeysDisallowedForRelyingParty(const WebCore::Securi
+
+void WebAuthenticatorCoordinatorProxy::signalAllAcceptedCredentials(const WebCore::SecurityOriginData&, WebCore::AllAcceptedCredentialsOptions&& options, CompletionHandler<void()>&& completionHandler)
+{
+    auto userHandle = base64URLDecode(options.userId);
```
error checking

nmahendru commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -1281,6 +1287,44 @@ static inline void getArePasskeysDisallowedForRelyingParty(const WebCore::Securi
+    auto userHandle = base64URLDecode(options.userId);
+    RetainPtr<NSMutableArray<NSData *>> credentialIds = adoptNS([[NSMutableArray alloc] init]);
+    for (auto& credential : options.allAcceptedCredentialIds) {
+        auto decodedCredentialId = base64URLDecode(credential);
```
error checking

nmahendru commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -1281,6 +1287,44 @@ static inline void getArePasskeysDisallowedForRelyingParty(const WebCore::Securi
+
+void WebAuthenticatorCoordinatorProxy::signalCurrentUserDetails(const WebCore::SecurityOriginData&, WebCore::CurrentUserDetailsOptions&& options, CompletionHandler<void()>&& completionHandler)
+{
+    auto userHandle = base64URLDecode(options.userId);
```
error checking

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/cbf75f2bdc3c28cf417bd044e4fc233088be9162. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/41630

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/62e856acc71c3a0ecdb8d99a1a4ece6862c3580c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/41630

nmahendru commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -1985,6 +1985,7 @@
 		A7F35F5F2B1959C000E43D98 /* RemoteLayerWithInProcessRenderingBackingStore.h in Headers */ = {isa = PBXBuildFile; fileRef = A7F35F5E2B1959B300E43D98 /* RemoteLayerWithInProcessRenderingBackingStore.h */; };
 		AAB145E6223F931200E489D8 /* PrefetchCache.h in Headers */ = {isa = PBXBuildFile; fileRef = AAB145E4223F931200E489D8 /* PrefetchCache.h */; };
 		AAFA634F234F7C6400FFA864 /* AsyncRevalidation.h in Headers */ = {isa = PBXBuildFile; fileRef = AAFA634E234F7C6300FFA864 /* AsyncRevalidation.h */; };
+		AE35BA1A2D70E6AC004DC511 /* CredentialUpdaterShim.h in Headers */ = {isa = PBXBuildFile; fileRef = AE35BA192D70E6AC004DC511 /* CredentialUpdaterShim.h */; };
```
This needs to go now I guess ?

nmahendru commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -14013,6 +14016,14 @@
+		AEB3CEDD2D72714000792466 /* Recovered References */ = {
+			isa = PBXGroup;
+			children = (
+				AE35BA192D70E6AC004DC511 /* CredentialUpdaterShim.h */,
```
Basically all references to CredentialUpdaterShim.h

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b6cf26fb44e57efeff2a3116a2ad48481ec926e8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/41630

nmahendru commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -57,6 +60,13 @@
+@interface CredentialUpdaterShim : NSObject
++ (void)signalUnknownCredentialWithRelyingPartyIdentifier:(NSString *)relyingPartyIdentifier credentialID:(NSData *)credentialID completionHandler:(void (^)(NSError *))completionHandler;
++ (void)signalAllAcceptedCredentialsWithRelyingPartyIdentifier:(NSString *)relyingPartyIdentifier userHandle:(NSData *)userHandle acceptedCredentialIDs:(NSArray<NSData *> *)acceptedCredentialIDs completionHandler:(void (^)(NSError *))completionHandler;
++ (void)signalCurrentUserDetailsWithRelyingPartyIdentifier:(NSString *)relyingPartyIdentifier userHandle:(NSData *)credentialID newName:(NSString *)newName completionHandler:(void (^)(NSError *))completionHandler;
```
WK_TBA_MACOS and friends here ?

nmahendru reviewed (approved):
Assuming the bots are happy 👍

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8dc08d845ef75a190b7858c15aff076fcabdc022. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/41630

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e994b05e5c93b268ca301d8a76cac8b03405413f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/41630

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4c5a61bdccf5ce281e04b3211b978b2a70f5251c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/41630

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0be7042b59c7b0c388cc441c05f8f5485f57fcfc. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/41630

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c74cde06bbc29d9c36da32d00ff599a0d10fc3e4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/41630

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b7201075b68efd974d18ff4af2ae9e382d2d3284. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/41630

pascoej committed:
Implement WebAuthn Signal API
https://bugs.webkit.org/show_bug.cgi?id=278339
rdar://problem/134285651

Reviewed by Nitin Mahendru.

This patch implements the new WebAuthn signal API. This allows RPs to
tell authenticators what credentials it will accept and also to update
credential information.

We don't currently mock the information nessesary to write tests for these
methods in LayoutTests.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/webauthn/AllAcceptedCredentialsOptions.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredential.idl.
* Source/WebCore/Modules/webauthn/AllAcceptedCredentialsOptions.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredential.idl.
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::signalUnknownCredential):
(WebCore::AuthenticatorCoordinator::signalAllAcceptedCredentials):
(WebCore::AuthenticatorCoordinator::signalCurrentUserDetails):
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.h:
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinatorClient.h:
* Source/WebCore/Modules/webauthn/CurrentUserDetailsOptions.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredential.idl.
* Source/WebCore/Modules/webauthn/CurrentUserDetailsOptions.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredential.idl.
* Source/WebCore/Modules/webauthn/PublicKeyCredential.cpp:
(WebCore::PublicKeyCredential::signalUnknownCredential):
(WebCore::PublicKeyCredential::signalAllAcceptedCredentials):
(WebCore::PublicKeyCredential::signalCurrentUserDetails):
* Source/WebCore/Modules/webauthn/PublicKeyCredential.h:
* Source/WebCore/Modules/webauthn/PublicKeyCredential.idl:
* Source/WebCore/Modules/webauthn/UnknownCredentialOptions.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredential.idl.
* Source/WebCore/Modules/webauthn/UnknownCredentialOptions.idl: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredential.idl.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.h:
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/AuthenticationServices/CredentialUpdaterShim.h: Copied from Source/WebCore/Modules/webauthn/PublicKeyCredential.idl.
* Source/WebKit/WebKitSwift/AuthenticationServices/CredentialUpdaterShim.swift: Added.
(CredentialUpdaterShim.signalUnknownCredential(_:credentialID:)):
(CredentialUpdaterShim.signalAllAcceptedCredentials(_:userHandle:acceptedCredentialIDs:)):
(CredentialUpdaterShim.signalCurrentUserDetails(_:userHandle:newName:)):
* Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.cpp:
(WebKit::WebAuthenticatorCoordinator::signalUnknownCredential):
(WebKit::WebAuthenticatorCoordinator::signalAllAcceptedCredentials):
(WebKit::WebAuthenticatorCoordinator::signalCurrentUserDetails):
* Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.h:

Canonical link: https://commits.webkit.org/291458@main
